File: cases/taint/csharp/function-field-negative/FunctionFieldNegative.cs
using System;

namespace DataFlowBench;

static class FunctionFieldNegative
{
    sealed class Holder
    {
        public Action<string> Fn = value => { };
    }

    static string dfb_source() // DFB-SOURCE: function-field-input
    {
        return "tainted";
    }

    static void dfb_sink(string value) { } // DFB-SINK: function-field-sink

    static void Invoke(Holder target, string value)
    {
        target.Fn(value);
    }

    static void Run()
    {
        Holder holder = new Holder();
        Holder otherHolder = new Holder();
        holder.Fn = value => // DFB-WITNESS: function-field-store
        {
            dfb_sink(value);
        };
        otherHolder.Fn = value =>
        {
            dfb_sink("clean");
        };
        Invoke(otherHolder, dfb_source());
    }
}

File: cases/taint/rust/function-field-negative/function_field_negative.rs
fn dfb_source() -> i32 { // DFB-SOURCE: function-field-input
    1
}

fn dfb_sink(value: i32) {} // DFB-SINK: function-field-sink

struct Holder {
    handler: Box<dyn Fn(i32)>,
}

fn invoke(target: &Holder, value: i32) {
    (target.handler)(value);
}

fn run() {
    let holder = Holder {
        handler: Box::new(|value| { // DFB-WITNESS: function-field-store
            dfb_sink(value);
        }),
    };
    let other_holder = Holder {
        handler: Box::new(|value| {
            dfb_sink(0);
        }),
    };
    invoke(&other_holder, dfb_source());
}
